server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,330 @@
+    }
+
+    // package private for testing
+    Path operatorSettingsDir() {
+        return operatorSettingsDir;
+    }
+
+    // package private for testing
+    Path operatorSettingsFile() {
+        return operatorSettingsDir().resolve(SETTINGS_FILE_NAME);
+    }
+
+    // platform independent way to tell if a file changed
+    // we compare the file modified timestamp, the absolute path (symlinks), and file id on the system
+    boolean watchedFileChanged(Path path) throws IOException {
+        if (Files.exists(path) == false) {
+            return false;
+        }
+
+        FileUpdateState previousUpdateState = fileUpdateState;
+
+        BasicFileAttributes attr = Files.readAttributes(path, BasicFileAttributes.class);
+        fileUpdateState = new FileUpdateState(attr.lastModifiedTime().toMillis(), path.toRealPath().toString(), attr.fileKey());
+
+        return (previousUpdateState == null || previousUpdateState.equals(fileUpdateState) == false);
+    }
+
+    @Override
+    protected void doStart() {
+        // We start the file watcher when we know we are master from a cluster state change notification.
+        // We need this additional flag, since cluster state can change after we've shutdown the service
+        // causing the watcher to start again.
+        this.active = true;
+    }
+
+    @Override
+    protected void doStop() {
+        this.active = false;
+        logger.debug("Stopping file settings service");
+        stopWatcher();
+    }
+
+    @Override
+    protected void doClose() {}
+
+    private boolean currentNodeMaster(ClusterState clusterState) {
+        return clusterState.nodes().getMasterNodeId().equals(clusterState.nodes().getLocalNodeId());
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        ClusterState clusterState = event.state();
+        setWatching(currentNodeMaster(clusterState), initialState);
+        initialState = false;
+    }
+
+    private void setWatching(boolean watching, boolean initialState) {
+        if (watching) {
+            startWatcher(initialState);
+        } else {
+            stopWatcher();
+        }
+    }
+
+    // package private for testing
+    boolean watching() {
+        return this.watchService != null;
+    }
+
+    synchronized void startWatcher(boolean onStartup) {
+        if (watching() || active == false) {
+            // already watching or inactive, nothing to do
+            return;
+        }
+
+        logger.info("starting file settings watcher ...");
+
+        Path settingsDir = operatorSettingsDir();
+
+        /*
+         * We essentially watch for two things:
+         *  - the creation of the operator directory (if it doesn't exist), symlink changes to the operator directory
+         *  - any changes to files inside the operator directory if it exists, filtering for settings.json
+         */
+        try {
+            this.watchService = PathUtils.getDefaultFileSystem().newWatchService();
```

Comment:
Rather than grab the default filesystem, it is probably better to grab it from the config dir?